it/src/main/java/org/corfudb/universe/universe/vm/VmUniverse.java

```diff
@@ -0,0 +1,319 @@
+
+        vms.putAll(deployment);
+    }
+
+    /**
+     * Deploy a {@link Group} on existing VMs according to input parameter.
+     */
+    @Override
+    protected Group deployGroup(GroupParams groupParams) {
+        switch (groupParams.getNodeType()) {
+            case CORFU_SERVER:
+                VmCorfuCluster cluster = VmCorfuCluster.builder()
+                        .universeParams(universeParams)
+                        .params(ClassUtils.cast(groupParams))
+                        .vms(ImmutableMap.copyOf(vms))
+                        .build();
+
+                return cluster.deploy();
+            case CORFU_CLIENT:
+                throw new UniverseException("Not implemented corfu client. Group config: " + groupParams);
+            default:
+                throw new UniverseException("Unknown node type");
+        }
+    }
+
+    /**
+     * Shutdown the {@link Universe} by stopping each of its {@link Group}.
+     */
+    @Override
+    public void shutdown() {
+        if (destroyed.getAndSet(true)) {
+            log.info("Can't shutdown vm universe. Already destroyed");
+            return;
+        }
+
+        log.info("Shutdown the universe: {}, params: {}", universeId, groups);
+        shutdownGroups();
+    }
+
+    @Override
+    public Universe add(GroupParams groupParams) {
+        throw new UnsupportedOperationException("Not implemented");
+    }
+
+    private CompletableFuture<VirtualMachine> deployVmAsync(String vmName) throws NodeException {
+        log.info("Deploy vm asynchronously: {}", vmName);
+        return CompletableFuture.supplyAsync(() -> deployVm(vmName), executor);
+    }
+
+    @Override
+    protected void shutdownGroups() {
+        super.shutdownGroups();
+
+        ImmutableSet<String> groupNames = universeParams.getGroups().keySet();
+
+        groupNames.forEach(groupName -> {
+            Group group = groups.get(groupName);
+            if (group != null) {
+                return;
+            }
+            universeParams
+                    .getGroupParams(groupName, GroupParams.class)
+                    .<VmServerParams>getNodesParams()
+                    .forEach(serverParams -> {
+                        CorfuServerOnVm node = CorfuServerOnVm.builder()
+                                .universeParams(universeParams)
+                                .params(serverParams)
+                                .vm(vms.get(serverParams.getVmName()))
+                                .build();
+
+                        try {
+                            node.destroy();
+                        } catch (NodeException ex) {
+                            log.info("Can't destroy corfu server: {}", groupName, ex);
+                        }
+                    });
+        });
+    }
+
+    /**
+     * Deploy and power on a VM appliance in vSphere.
+     *
+     * @return VirtualMachine instance
+     */
+    private VirtualMachine deployVm(String vmName) throws NodeException {
```

Comment:
We have 2 methods - deplyVM and vmDeploy.
Can we come up with better names?

Reply:
Fixed (see ApplianceManager)